openedx_events/bridge/avro_attrs_bridge.py

```diff
@@ -1,253 +1,231 @@
+        for extended_class, extension in self.extensions.items():
+            if issubclass(type(value), extended_class):
+                return extension.serialize(value)
+        # Case 3: Default
+        return value
+
+    def _serialize_non_attrs_instance(self, _, field, value):
         """
-        Pass this callback into attrs.asdict function as "value_serializer" arg.
+        Use an extension to serialize a value of the appropriate class.
 
-        Serializes values for which an extention exists in self.extensions dict.
+        Used as a callback to attr.asdict to handle any inner fields that are not attrs classes or primitives
         """
         extension = self.extensions.get(field.type, None)
         if extension is not None:
             return extension.serialize(value)
         return value
 
-    def deserialize(self, data: bytes, writer_schema=None) -> object:
+    def from_dict(self, data: dict):
         """
-        Deserialize data into self.attrs_cls instance.
+        Convert dict into event data that can be sent with self._signal.
+
+        Arguments:
+            data: Dictionary returned from AvroDeserializer
 
-        Args:
-            data: bytes that you want to deserialize
-            writer_schema: pass the schema used to serialize data if it is differnt from current schema
+        Returns:
+            dict: Event data dictionary
         """
-        data_file = io.BytesIO(data)
-        if writer_schema is not None:
-            record = fastavro.schemaless_reader(
-                data_file, writer_schema, self.schema_dict
-            )
-        else:
-            record = fastavro.schemaless_reader(data_file, self.schema_dict)
-        return self.dict_to_attrs(record["data"], self._attrs_cls)
+        return dict([(data_key, self._deserialized_avro_dict_to_object(data[data_key], data_type))
+                     for data_key, data_type in self._signal.init_data.items()])
 
-    def dict_to_attrs(self, data: dict, attrs_cls):
+    def _deserialized_avro_dict_to_object(self, data: dict, data_type):
         """
-        Convert data into instantiated object of attrs_cls.
+        Convert dictionary entry into an instance of data_type.
+
+        Used to convert messages from an AvroDeserializer into events that can be sent by the
+        appropriate signal instance
+
+        Arguments:
+            data: Dictionary returned from AvroDeserializer
+            data_type: Desired Python data type, eg `str`, `CourseKey`, `CourseEnrollmentData`
+
+        Returns:
+            An instance of data_type
         """
-        for attribute in attrs_cls.__attrs_attrs__:
+        if not hasattr(data_type, '__attrs_attrs__'):
+            if extension := self.extensions.get(data_type, None):
+                return extension.deserialize(data)
+            else:
+                raise TypeError(
+                    f"Unable to deserialize {data_type} data, please add extension for custom data type"
```

Comment:
Note: It may be useful to vary the sentence slightly to indicate where the error happened without requiring a stacktrace.